File: SeongYeon/ts/leetcode/20_ValidParentheses.ts
// Runtime: 68ms(Beats 59.24%)
// Memory: 44.89mb(Beats 43.19%)

function isValid(s: string): boolean {
  const stack = [];
  const bracketPair = {
    ')': '(',
    '}': '{',
    ']': '[',
  };

  for (const bracket of s) {
    if (bracket in bracketPair) {
      if (!stack.length || stack.pop() !== bracketPair[bracket]) return false;
    } else stack.push(bracket);
  }

  return !stack.length;
}

File: JiHwan/leetcode/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        gwalho = {'{': '}', '(': ')', '[': ']'}
        for i in s:
            if i in gwalho:
                stack.append(i)
            else:
                if not stack or gwalho[stack.pop()] != i:
                    return False

        if stack:
            return False
        else:
            return True

